Use `blockwise_broadcast_reduce` in reduction fusions.

manupak opened the pull request:
~~[TODO] : add more e2e tests and unit tests for passes.~~

Currently, we handle reductions that are being fused with gemm-like operations
by using atomic stores to the destination buffer. This can be cripplingly slow when
most of the output is being reduced as evidenced in layer_norm cases.

This PR adds the ability to `blockwise_broadcast_reduce` on the block sub-tiles of 
of gemm output.
However, in-order to that we need to make sure the reduction dimension is uniformly
distributed across the blocks. This is achieved by : 
1) Firstly, this PR introduces a utility where for a given set of upper dimensions, it can
    traverse a transform stack and produce a list of sub-dimensions per each lower dimension
    where the upper reduction axes are mapped to.
2) Then, this PR introduces `ShuffleGemmForReductions` pass, which will split and transpose
    the parallel dimension of the gemm such that reduction dimension is uniformly
    distributed across the blocks.
3) Then at AlignTiling pass, we extract the block subtile when fusing in the `rock.reduce` operator. 
    Then perform a `blockwise_broadcast_reduce` on the block subtile.
4) Since we only want to write the partial reductions per block, we pad out broadcasted part of the subtile.
    (We rely on any block coordinate that goes to the padded region within the block will not be written out)
5) Then we need to do `Recombine` the modified sub-tile coordinate maps with grid-only coordinates maps.
     a) Here, we drop all the upper-dimensions except `g_block`, `m_block` and `n_block` and obtain the grid-only transform map 
         stack.
     b) In parallel, we re-use getLowerSubDimensions utility to figure out which sub-dimension gets mapped with the above 
         grid-only dimensions.
     c) Then we extract of those sub-dimensions in a bottom up fashion and stitch it up with the said  grid-only transform map 
         stack.

I ll try to create some slides to explain all these. 

In the cases I ve tests, this yields two orders of magnitude (~100x) gains over the pure atomics approach of doing reduction fusions.

manupak committed:
Initial code upto insertion of blockwise_reduce
TODO : from creating a new view to represent
partial reduction in a global view.

manupak committed:
Merge branch 'develop' of github.com:ROCm/rocMLIR into fuse-reduction-blockwise-reduce

manupak committed:
Added code for align tiling to insert
a blockwise_reduction to reduce the number
of atomic_add s issued.

manupak committed:
* Moved blockwise to threadwise lowering after linalg align
* fixed blocksize

manupak committed:
* add a e2e mixr test

manupak committed:
Its not actually working...
There is a fractional dim being create due to reduction
being applied on a partial dimensions

manupak committed:
* add debug prints under LLVM_DEBUG()
* add asserts to where fractional dims are created when removing
  upperDims

manupak committed:
Added support for non-removed subdims + padding
TODO : support cases where subdim is removed in pad

manupak committed:
Added support removeSubDim in padded dim + tests

manupak committed:
* clang format

manupak committed:
Merge branch 'padding-for-remove-upperdims' of github.com:manupak/rocMLIR into fuse-reduction-blockwise-reduce

manupak committed:
Add pad reductions fusions pass

* Currently it has has code obtain all
  views leading to gemm.

manupak committed:
* add a utility to trace a dimension through a tr stack
  back to sub dimensions from upper to lower.

manupak committed:
WIP trying resplit reduction dimension into gcd d
with DPerBlock

manupak committed:
Added input reordering

manupak committed:
Input tranposes are done.
Next: output transposes

manupak committed:
tranposing pass runs..
but I still see blocking non dividability.
me need re-think transposes

manupak committed:
* test case working

manupak committed:
* remove printing

manupak committed:
* tuning loop runs for that test case

manupak committed:
* tuning loop really runs for the test case
* abort if reduce to gemm is not invertible
* return failure for removeUPperDims divisbility cases
  so that blockwise reductions are not used.

manupak committed:
*clean up
* make th shuffle pass run on largest reduction

manupak committed:
Further clean up
renamed pass
TODO : fix recombine logic

manupak committed:
* fixed recombine logic
* added another iface for getLowerSubDIms

manupak committed:
* handle tail subdims

manupak committed:
* make the test case check for all outputs

manupak committed:
* check-rocmlir works

manupak committed:
Merge branch 'develop' of github.com:ROCm/rocMLIR into fuse-reduction-blockwise-reduce

manupak committed:
*clang-format

manupak committed:
* fixed the bug that we should only be changing the gemm output
  with additional views

manupak committed:
* add logic to avoid padding cases for reduction fusions

manupak committed:
* clang format

manupak committed:
* update unit tests

krzysz00 commented on mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:
```diff
@@ -151,15 +151,22 @@ void rock::buildKernelPipeline(OpPassManager &pm,
+  // We want to delay blockwise lowering in the fusion cases
+  // until after linalg align pass because with reduction fusion
+  // it may introduce blockwise_reductions.
+  if (!options.enableFusion) {
```
Nit: we could move `rock-pipeline` and friends out of the `enable-fusion` if statement, so that we can just move `-blockwise-gemm-to-threadwise` after `-linalg-align`

manupak replied:
Having read the code again, here we have `rock-pipeline` in enableFusion!
Hence, I believe we dont do `enableFusion` for non-fusion cases -- its kind of broken isn;t it? 

In retrospect, do we really need to have `enableFusion` ? The fact that nothing breaks, indicates its not used. WDYT ? 
If agreed, we can remove `enableFusion` option in a followup ?

krzysz00 replied:
Yeah, let's lose `enableFusion`. I think it was added so the MIOpen people wouldn't see the overhead

manupak replied:
Just to keep the bisectability, I ll do that in a seperate PR if thats ok..

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -405,8 +406,10 @@ void LinalgAlignRewriter::notifyMatchFailure(
                               Value dest, ArrayRef<TransformMapAttr> views) {
-  for (TransformMapAttr trMap : llvm::reverse(views)) {
-    dest = rewriter.create<TransformOp>(loc, dest, trMap);
+  if (!views.empty()) {
```
Why this if statement? Shouldn't the loop be trivial if views is empty?

manupak replied:
removed

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+  std::tie(dest, destTrs, std::ignore) =
+      untransform(rewriter, threadwiseWriteOp.getDest());
+  ArrayAttr toBeReducedViews = prependUpperViews(rewriter, extraViews, destTrs);
+  TransformMapAttr firstCoordTransform =
```
`return notifyMatchFailure()` or some such if there are no transforms at all so we don't crash here?

manupak replied:
extraViews can never be empty by definition because those correspond to g_block x m_block x n_block x tid x iter -> G X M X N. (though destTrs could be empty, thus concat of the two will also never be empty)

krzysz00 replied:
In the future, we might get rid of `extraViews`, for example

manupak replied:
that would mean they would go to destTrs.. right? 
coz somewhere there needs to hold the maps that starts with g_block x m_block x n_block x tid x iter -> G X M X N
So the concat would still be not empty.. 

(I can put an assert)

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+  ArrayAttr toBeReducedViews = prependUpperViews(rewriter, extraViews, destTrs);
+  TransformMapAttr firstCoordTransform =
+      cast<TransformMapAttr>(toBeReducedViews[0]);
+  int upperRank = firstCoordTransform.getUpperBounds().size();
```
This might want to be an `int64_t` to match the others

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+
+  // Extract grid-only dims
+  removeIndicesSet.clear();
+  for (int64_t i = 3; i < upperRank; i++) {
```
Why `3` not `upperRank - 2`?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+      TopDownTMBuilder toReducedView(rewriter, nameRefs, blockTileShape);
+      for (unsigned i = 0; i < blockTileShape.size(); i++) {
+        if (blockReductionAxis == i) {
+          toReducedView.pad({nameRefs[i]}, {0, blockTileShape[i] - 1});
```
... hold on, is this using `Pad{}` to take a length-1 dimension to a length-`blockTileShape[i]` one?

This feels like it could be `Broadcast{}`?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+        TransformMapAttr trMapAttr = cast<TransformMapAttr>(attr);
+        SmallVector<TransformAttr> trAttrs;
+        SmallVector<int64_t> gridUpperShape;
+        SmallVector<int64_t> gridLowerShape;
```
Do these need to be vectors?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+        llvm::SmallDenseMap<int64_t, SmallVector<unsigned>> upperDims;
+        int64_t dimInsertionPoint = 0;
+        // The three dimensions correspond to G, M and N here.
+        for (unsigned dim = 0; dim < 3; dim++) {
```
Same note re hardcoded 3

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,595 @@
+ArrayAttr getAllViewsFromSource(OpOperand *operand) {
+  Value val = operand->get();
+  SmallVector<Attribute> attrs;
+  while (TransformOp trOp = dyn_cast<TransformOp>(val.getDefiningOp())) {
```
This is `untransform()` no?

krzysz00 commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -264,13 +299,16 @@ void createGemmTuningRangeBF(TuningParamSet *newSpace,
                           (kind == TuningParamSetKind::Exhaustive ||
-                           succeeded(
-                               tuningInfo.couldBePerformant(info, gemmParams))))
+                           !nonExhaustiveSkipCondition))
```
Could you implement this into `couldBePerformant()` please? Maybe extend that function to take a `bool isReduction`?

krzysz00 replied:
... This is also, IMO, adding a `has_reduction` to the problem key, so that MIGraphX doesn't cache reduction and non-reduction GEMMs together

manupak replied:
I moved into couldBePerformant() and I d defer problem key change to be in a different PR for bisectability.

krzysz00 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2374,148 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+    return failure();
+  }
+  return subDimInfo;
+}
```
Out of curiosity, could this SubDimInfo stuff be used to simplify `collapseContiguousMerges()`?

(Not for this PR, just checking my understanding)

manupak replied:
potentially yes...

krzysz00 reviewed (commented):
First round of looking:
1. I've got the inevitable stash of nitpicks
2. The sub-dim gathering utility looks correct, and I have a solid amount of trust in `removeUpperDims()` at this point
3. The rats' nest that is constructing and manipulating all the transforms is calling out for some refactoring and utilities ... but that's not this PR

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -2033,7 +2033,8 @@ LogicalResult BlockwiseBroadcastReduceOp::verify() {
     return emitError(
-        "workspace should be at least the size of elements per block");
+               "workspace should be at least the size of elements per block ")
+           << blockwiseInputPartialReductionTensorElements << " " << wsShape[0];
```
print variable names as well? or remove if this was for debugging purposes?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,595 @@
+  func.walk([&](ReduceOp rOp) -> WalkResult {
+    TypedValue<ShapedType> rIn = rOp.getIn();
+    int64_t reduceDimSize =
+        rIn.getType().getShape()[rOp.getAxis().getZExtValue()];
```
could the type be vector? in that case we might want to multiply by the vector size as well

manupak replied:
we will not have vectors for rock.reduce op as elements.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,595 @@
+  if (largestReductionOp) {
+    auto &bufferDeps = getAnalysis<BufferDependencyAnalysis>();
+    rearrangeGemmParallelDimsForReduction(largestReductionOp, bufferDeps);
+  }
```
suggestion: add a debug message if the pass doesn't run

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,595 @@
+    LLVM_DEBUG(llvm::dbgs() << "gemmToReduceViews=" << views << "\n");
+    FailureOr<MNPerBlock> mnPerBlock = getMNPerBlock(gemmOp);
+    if (failed(mnPerBlock)) {
+      return;
```
suggestion: return failure() and print debug message from the caller (the pass will not run).

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2374,148 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+mlir::rock::getLowerSubDimensions(OpBuilder &b, ArrayAttr transformAttrs,
+                                  ArrayRef<int64_t> dims) {
+  llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> subDimInfo;
+  TransformMapAttr topMap = cast<TransformMapAttr>(transformAttrs[0]);
```
check transformAttrs is not empty

manupak committed:
Remove the hardcoding that reduction space being
3 dimensional.

manupak committed:
* remove further hardcodes

krzysz00 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2374,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+        } break;
+        case TransformType::Merge: {
+          SmallVector<int64_t> subDimStrides = getStrides(trAttr.getParams());
+          for (size_t subDim = 0; subDim < trAttr.getParams().size();
```
This could be a `llvm::zip()` or an `llvm::enumerate()` if you do need the index

manupak replied:
thanks! it did

krzysz00 reviewed (commented):
Very tiny note that might generalize

manupak committed:
Add support for AddDim & ConstDim
in getLowerSubDimensions

manupak committed:
add dot + reshape + reduce case

manupak committed:
Address feedback : fold fused reduction
skipping logic into couldBePerformantAPI.

manupak committed:
Merge branch 'develop' of https://github.com/ROCm/rocMLIR into fuse-reduction-blockwise-reduce

manupak committed:
* clang format

manupak committed:
Address review feedback : AlignTiling

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1145,21 +1146,406 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+          SmallVector<int64_t> splitSizes;
+          int64_t currSize = toBeReducedShape[dim];
+          for (const SubDimInfo &subDim : subDims) {
+            assert(currSize % (subDim.size * subDim.stride) == 0);
```
Why assert? Maybe fail here?

manupak replied:
refactored into a function and a failure return

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,597 @@
+      LLVM_DEBUG(llvm::dbgs()
+                 << "readOperand = " << readOperand->get() << "\n");
+      // Test against the write operand to guard against [MemRead, MemWrite]
+      if (readOperand && readOperand != writerOperand &&
```
Doesn't `BufferDependencyAnalysis` do this?

krzysz00 replied:
Or have this cached?

manupak replied:
This is heavily inspired by your code here : https://github.com/ROCm/rocMLIR/blob/82378ac44b8627ecbaf9078353fb53588090fe28/mlir/lib/Dialect/Rock/Transforms/Regularize.cpp#L282-L290.

I think its safer to have here as well.

krzysz00 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2066,32 +2075,61 @@ static FailureOr<rock::TransformMapAttr> removeUpperDimsFromMap(
+                      << "non divisible subDim removal found. aborting..\n");
+                  return failure();
+                }
+                // assert(maxStrideSubDim % rhsBoundForRemoval == 0);
```
Stray comment

krzysz00 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2374,172 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+    subDimInfo[dim].push_back({topMap.getUpperBounds()[dim], 1});
+  }
+
+  for (Attribute attr : transformAttrs) {
```
You can do a `.getAsRange<TransformMapAttr>()`

manupak committed:
Refactor blokwise_reduction insertions into a function
in aligntiling -- and allow it fail rather than having
asserts if something is as not expected.

manupak committed:
* clang format

manupak committed:
use untransform

manupak committed:
Address review feedback

manupak commented:
Thanks @krzysz00 for all the reviews on this big PR. I really appreciate it.
I have addressed all of them

I have not added the unit tests for passes yet and aling-tiling yet... which I will do next.
@dhernandez0 would you be able to take a look here ?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+  SetVector<int64_t> removeIndicesSet;
+  // We only want to keep tid x iter in the maps
+  // which is the last two for block subtile
+  for (int64_t i = 0; i < upperRank - 2; i++) {
```
I guess rank is always at least 2? tid, iter?

manupak replied:
yes... they must be there.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+  // We only want to keep tid in the maps
+  // which is the last two for block subtile tid
+  removeIndicesSet.clear();
+  for (int64_t i = 0; i < upperRank; i++) {
```
we can skip the loop and just add iter:
removeIndicesSet.insert(upperRank-1);
The rest of the indices are already inserted

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+  // We only want to keep iter in the maps
+  // which is the last one.
+  removeIndicesSet.clear();
+  for (int64_t i = 0; i < upperRank - 1; i++) {
```
same here, just remove upperRank-1 and add upperRank-2

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+
+  FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>>
+      lowerSubDims =
+          getLowerSubDimensions(rewriter, toBeReducedViews, {0, 1, 2});
```
should we iterate until upperRank-2 as in the previous code?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+    if (idx == (size_t)blockReductionAxis) {
+      ldsWorkspaceSize *= partialReductionsPerThread;
+    } else {
+      ldsWorkspaceSize *= size;
```
we could assert that memory required is not bigger than GPU LDS here. Reuse LDS pass would do it anyway, but it might be easier to debug if we add an assert here if this issue comes up.

manupak replied:
a better solution might be we add that to the verifier of `rock.gpu_alloc` ?

manupak replied:
ended up adding here as we want this merged before release.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+      reduceOp.getReduceMethodAttr(), blockSubTileViews.value(),
+      blockSubTileTidSliceViews.value(), threadSubTileViews.value(),
+      /*extraViews=*/nullptr,
+      getBlockSize(reduceOp->getParentOfType<func::FuncOp>()).value());
```
could we get this internally in the BlockwiseBroadcastReduceOp lowering instead?

manupak replied:
yep a good suggestion.. but I d do that seperately as a NFC change for bisectability reasons.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+
+  ViewLikeOpInterface viewOp =
+      ldsWorkspace.getDefiningOp<ViewLikeOpInterface>();
+  rewriter.create<GpuDeallocOp>(loc, viewOp.getViewSource());
```
nit: viewOp.getViewSource() -> ldsWorkspace

manupak replied:
no right... ldsWorkspace is view op with the right element type.. here we want the raw i8 memref.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2371,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+  TransformMapAttr topMap = cast<TransformMapAttr>(transformAttrs[0]);
+  for (int64_t dim : dims) {
+    // No point of tracing size 1 dimensions
+    if (topMap.getUpperBounds()[dim] == 1) {
```
could this cause issues if the code calling getLowerSubDimensions expects the dictionary to have this dim info?

manupak replied:
caller cant have a such an expectation.
it is an information provided by this util whether it contains or not.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2371,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+          for (size_t subDim = 0; subDim < trAttr.getParams().size();
+               subDim++) {
+            int64_t upperDim = trAttr.getUpperDims()[subDim];
+            if (currSubDimInfo.contains(upperDim)) {
```
I thought it currSubDimInfo.contains(upperDim) should always be true, when can this be false?

manupak replied:
as I said before, not all dimensions need to contain the interested subdimensions that we are tracing.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:
```diff
@@ -2319,3 +2371,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+          SmallVector<int64_t> subDimStrides = getStrides(trAttr.getParams());
+          for (auto [lowDim, subDimStride, param] : llvm::zip(
+                   trAttr.getLowerDims(), subDimStrides, trAttr.getParams())) {
+            int64_t upperDim = trAttr.getUpperDims()[0];
```
nit: move out of the loop

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+    ArrayRef<int64_t> blockTileShape = getLowerShape(blockSubTileViews.value());
+    SmallVector<SmallString<8>> names;
+    SmallVector<StringRef> nameRefs;
+    for (size_t i = 0; i < blockTileShape.size(); i++) {
```
nit: llvm::enumerate()

manupak replied:
I dont need the element from blockTileShape. 
I just need the idx. 
enumerate will provide both here.

manupak replied:
I ll remove the whole loop as I ve created a utility to do this now.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+    reduceOut = cast<TypedValue<ShapedType>>(
+        applyViewsOnDest(rewriter, loc, reduceOut, transformAttrs));
+    threadwiseWriteOp.getDestMutable().assign(reduceOut);
+    threadwiseWriteOp.setStoreMethodAttr(stMethod);
```
is there any case where we could get rid of atomics? maybe leave a TODO as a possible optimization if that's the case.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+        rewriter, blockSubTileViews.value(), arrayTransformAttrs);
+  }
+
+  // Recombine block dimensions
```
could you explain a bit more here? 
nit: consider moving it to a function

manupak replied:
I ve refactored recombine into a new function + docs.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+          TransformMapAttr::get(trAttrs, upperBounds, lowerBounds);
+      transformAttrs.push_back(newTrMap);
+    }
+    ArrayRef<int64_t> currLowerShape =
```
rock::getLowerShape

manupak replied:
no its not a ArrayAttr (yet) to use that

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+      TopDownTMBuilder toAddMissingBlockDims(rewriter, nameRefs,
+                                             currLowerShape);
+      {
+        toAddMissingBlockDims.passThrough({0, 1, 2}, {0, 1, 2});
```
could we do 0 to upperRank - 2?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+      int64_t dimInsertionPoint = 0;
+      for (unsigned dim = 0; dim < toBeReducedShape.size(); dim++) {
+        // The lower subDims contain sub-dimensions where blocking
+        // indices -- namesly g_block, m_block and n_block -- maps to
```
namesly -> namely

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+    }
+    LLVM_DEBUG(llvm::dbgs() << "lastMerge=" << lastMerge << "\n");
+    // The above view contains splitted sub-dims that are either associated
+    // with grid and non-grid dimensions. Then, we concat them into 6
```
"6 dimensions", I think 6 is not hardcoded here

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:
```diff
@@ -211,12 +215,41 @@ PopulateParams::paramsProbablyValid(OpBuilder &b,
+  // 16 is practically lowest m in MFMAs/WMMAs
+  // that could be performant. If the gemm sizes
+  // are not divisible by that, then we definitely
+  // need padding. Therefore, it cant use blockwise
```
cant -> can't

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:
```diff
@@ -211,12 +215,41 @@ PopulateParams::paramsProbablyValid(OpBuilder &b,
+  // reductions. Thus, it becomes a competition among
+  // atomic_store based reduction kernels.
+  if (gemmSize.m % 16 != 0) {
+    return success();
```
shouldn't this be failure()?

manupak replied:
no ... 
If the size is not divisible by 16 then it means all the configs will have gemm padding for the tile sizes we support.
Thus, the tuning is done among atomic based solutions.  Therefore, all the solutions must be tuned.

manupak replied:
added more docs as discussed.

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:
```diff
@@ -211,12 +215,41 @@ PopulateParams::paramsProbablyValid(OpBuilder &b,
+    return success();
+  }
+  if (gemmSize.n % 16 != 0) {
+    return success();
```
same here

manupak replied:
see the reply above

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:
```diff
@@ -211,12 +215,41 @@ PopulateParams::paramsProbablyValid(OpBuilder &b,
   return populateDerived(params);
 }
 
+static LogicalResult couldFusedReductionBePerformant(const GemmSize &gemmSize,
```
nit: I'd say this is not a failure/success function but a bool type

manupak replied:
im being consistent with "couldBePerformant" API.

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -364,7 +364,7 @@ void createQuickTuningRange(TuningParamSet *newSpace,
                                             info.gemmBType, info.arch),
              info.gemmSize)) {
-      if (succeeded(tuningInfo.paramsProbablyValid(b, info, param)))
+      if (succeeded(tuningInfo.couldBePerformant(info, param)))
```
should we check if params are valid anyway?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+static void
+rearrangeGemmParallelDimsForReduction(ReduceOp rOp,
+                                      const BufferDependencyAnalysis &deps) {
+  FailureOr<std::tuple<ArrayAttr, Operation *>> res =
```
nit: res -> reduceInfo? or some other more descriptive variable name

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+      return;
+    }
+    FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<mlir::rock::SubDimInfo>>>
+        subDimensions = getLowerSubDimensions(rewriter, invertedViews, 2);
```
use a variable name for dimension to make it clear what it is

dhernandez0 replied:
also, use reduction dim from reduceop

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+    llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> reductionSubDims) {
+  // Split the reduction and non-reduction splits
+  int64_t totalReductionSizeM = 1;
+  for (const SubDimInfo &sdInfo : reductionSubDims[1]) {
```
we should not assume reductionSubDims contains 1 or 2

dhernandez0 replied:
also, should we also consider G? for batch norm?

manupak replied:
We only shuffle dimensions where we do tiling and G is not a tiling dimension.
I ll add a comment.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+        mnPerBlock.value().NPerBlock, subDimensions.value());
+    rewriter.setInsertionPointAfterValue(gemmOut);
+    Value trGemmOut = gemmOut;
+    Value firstUse = nullptr;
```
I don't think this is used anywhere

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+    SmallVector<SubDimInfo> nReductionSubDimInfo = reductionSubDims[2];
+
+    unsigned dimInsertionPoint = 1;
+    {
```
instead of doing it for m, then n. We could make it generic and iterate over m and n (and g if we consider doing that as well).

manupak replied:
I prefer to keep this very specific to M & N not be general because it relies on things like MPerBlock and NPerBlock present in the gridwise_gemm_accel op.

Also, it would be helpful, to see them in IR which dims are being split explicitly as "M" or "N".

In principle I agree a general solution might work here, but I d delay that until this is mature enough. 
So that it allows me to inspect easily to handle bugs. WDYT ?

dhernandez0 replied:
Fair enough, let's keep it like this. In the future we could make it generic but still write M/N in the IR, hardcoding dimStr = {"m", "n"}.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+    toRecombineMN.getStartNames(startNames);
+
+    // M
+    {
```
nit: same here, make it generic and iterate over dims?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+  ArrayAttr additionalViewsA = builder.getArrayAttr({});
+  if (!reductionSubDims[1].empty()) {
+    additionalViewsA = reorderReductionDims(
+        toReductionSplitA, reductionSubDims[1], "M", m, mPerBlock);
```
reductionSubDims might not contain 1

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp:
```diff
@@ -0,0 +1,594 @@
+  ArrayAttr additionalViewsB = builder.getArrayAttr({});
+  if (!reductionSubDims[2].empty()) {
+    additionalViewsB = reorderReductionDims(
+        toReductionSplitB, reductionSubDims[2], "N", n, nPerBlock);
```
reductionSubDims might not contain 2

manupak committed:
Address Daniel's feedback except the recombine
refactor

manupak committed:
clang format

manupak committed:
Add more docs for could be performant

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:
```diff
@@ -1332,25 +1334,23 @@ static LogicalResult insertBlockwiseReduction(
+        for (unsigned i = 0; i < upperRank - 2; i++) {
+          gridOnlyDimIdxs.push_back(i);
+        }
+        toAddMissingBlockDims.passThrough(gridOnlyDimIdxs, {0, 1, 2});
```
toAddMissingBlockDims.passThrough(gridOnlyDimIdxs, gridOnlyDimIdxs);

manupak replied:
my bad.

manupak committed:
* further address comments

dhernandez0 commented on mlir/test/fusion/pr-e2e/mixr-multi-reduce-mo-4d.mlir:
```diff
@@ -0,0 +1,29 @@
+    %11 = migraphx.mul %4, %4 : <2x32x10x64x64xf32, 1310720x40960x4096x64x1>, <2x32x10x64x64xf32, 1310720x40960x4096x64x1> -> <2x32x10x64x64xf32, 1310720x40960x4096x64x1>
+    %12 = migraphx.mul %11, %10 : <2x32x10x64x64xf32, 1310720x40960x4096x64x1>, <2x32x10x64x64xf32, 0x0x0x0x0> -> <2x32x10x64x64xf32, 1310720x40960x4096x64x1>
+    %13 = migraphx.reshape %12 {dims = [2, 32, 40960]} : <2x32x10x64x64xf32, 1310720x40960x4096x64x1> -> <2x32x40960xf32, 1310720x40960x1>
+    %14 = migraphx.reduce_sum %13 {axes = [2]} : <2x32x40960xf32, 1310720x40960x1> -> <2x32x1xf32, 32x1x1>
```
nit: add tests where the axis of reduction is not 2. Also, tests where m/n is not reduced at all. This is to test the fixes introduced today.

manupak replied:
Thanks. will add them.
I also need to fix/add unit test for align-tiling/rock-shuffle-gemm-for-reductions

manupak committed:
Refactor recombine step into a new function
with docs

manupak committed:
* add align tiling unit tests

manupak committed:
Add reduction axis 1 test case
Add n dimension only reduction test case

manupak committed:
* add unit test for shuffle gemm pass

manupak committed:
* fix missing new lines

manupak commented:
@dhernandez0 I ve added the promised tests now!

dhernandez0 commented:
> @dhernandez0 I ve added the promised tests now!

Thanks! I was wondering if we should add a test where the reduction occurs along the G axis only for completeness. I understand it would only involve atomics, but it might be good to include it just in case. I've already approved the PR, so whatever you decide.
